presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java

```diff
@@ -145,10 +161,48 @@ public RuleAssert matches(PlanMatchPattern pattern)
         return this;
     }
 
+    public <T extends Trait> RuleAssert satisfies(T trait)
+    {
+        RuleApplication ruleApplication = applyRule();
+        if (!ruleApplication.satisfies(trait)) {
+            Optional<T> actual = ruleApplication.lookup.resolveTrait(ruleApplication.planNode, trait.getType());
+            fail(String.format("Plan node does not satisfy trait: %s, but it has: %s", trait, actual));
+        }
+        return this;
+    }
+
+    public <T extends Trait> RuleAssert doesNotSatisfy(T trait)
+    {
+        RuleApplication ruleApplication = applyRule();
+        if (ruleApplication.satisfies(trait)) {
+            Optional<T> actual = ruleApplication.lookup.resolveTrait(ruleApplication.planNode, trait.getType());
+            fail(String.format("Expected plan node to not satisfy trait: %s, but it has: %s", trait, actual.get()));
+        }
+        return this;
+    }
+
+    public <T extends Trait> RuleAssert hasNo(TraitType<T> traitType)
+    {
+        RuleApplication ruleApplication = applyRule();
+        Optional<T> trait = ruleApplication.lookup.resolveTrait(ruleApplication.planNode, traitType);
+        if (trait.isPresent()) {
+            fail(String.format("Expected plan node to not have trait, but found: %s", trait.get()));
+        }
+        return this;
+    }
+
     private RuleApplication applyRule()
     {
         SymbolAllocator symbolAllocator = new SymbolAllocator(symbols);
-        Memo memo = new Memo(idAllocator, plan);
+        Memo memo;
+        if (beforeRules.isEmpty()) {
+            memo = new Memo(idAllocator, plan);
+        }
+        else {
+            IterativeOptimizer iterativeOptimizer = new IterativeOptimizer(new StatsRecorder(), beforeRules);
+            memo = iterativeOptimizer.explore(plan, session, symbolAllocator, idAllocator);
```

Comment:
I really am not a fan of this. You either use a `memo` you constructed _or_ you use some unknown `memo` instance, a detail of `Iterative`. Maybe we should be testing `Iterative(beforerules + the tested rule)` as a unit instead?

Reply:
I try to address the issue more holistically in: https://github.com/Teradata/presto/tree/ks/rule_apply_iterative.
Currently `RuleAssert` is rather poor since it uses different mechanism for rule application than the optimizer itself. For instance it doesn't really test if rule ends.